Clarify path validation and connection migration (#4102)
quicwg/base-drafts: draft-ietf-quic-transport.md

## Path Validation {#migrate-validate}

Path validation is used by both peers during connection migration
(see {{migration}}) to verify reachability after a change of address.
In path validation, endpoints test reachability between a specific local
address and a specific peer address, where an address is the two-tuple of
IP address and port.

> Path validation tests that packets sent on a path to a peer are received by that
> peer. Path validation is used to ensure that packets received

Author: ```suggestion
Path validation tests that packets sent on a path to a peer are
received by that peer. Path validation is used to ensure that packets received
```

> peer. Path validation ensures that packets received from a migrating peer do
> not carry a spoofed source address.

Reviewer: I personally found `is used to ensure` clearer than `ensures`

```suggestion
peer. Path validation is used to ensure that packets received
from a migrating peer do not carry a spoofed source address.
```

Path validation does not validate that a peer can send in the return direction.
Acknowledgments cannot be used for path validation as they contain insufficient

> entropy and might be spoofed. No method is provided to establish return
> reachability, as endpoints independently determine reachability on each
> direction of a path.

Reviewer: I found `No method is provided to establish return reachability` confusing. How about:

```suggestion
entropy and might be spoofed. Endpoints independently determine reachability
on each direction of a path, and therefore return reachability can only be
established by the peer.
```

traversal needs additional synchronization mechanisms that are not provided
here.

An endpoint MAY include other frames with the PATH_CHALLENGE and PATH_RESPONSE
frames used for path validation.  In particular, an endpoint can include
PADDING with a PATH_CHALLENGE for Path Maximum Transfer Unit (PMTU) discovery
(see {{pmtud}}), and/or include a PATH_RESPONSE with its own PATH_CHALLENGE.

> When probing a new path, an endpoint might want to ensure that its peer has an

Author: This paragraph is very duplicative of a paragraph in "Probing a new path" under connection migration, so I think we should drop one of them, but I'm not clear on which.

Reviewer 2: The latter can probably go.

Author: Done.

traversal needs additional synchronization mechanisms that are not provided
here.

An endpoint MAY include other frames with the PATH_CHALLENGE and PATH_RESPONSE
frames used for path validation.  In particular, an endpoint can include
PADDING with a PATH_CHALLENGE for Path Maximum Transfer Unit (PMTU) discovery

> (see {{pmtud}}), and/or include a PATH_RESPONSE with its own PATH_CHALLENGE.

Reviewer: This last bit sounded backwards to me in the original text. I think as a server I expect to commonly receive an unprompted PATH_CHALLENGE, and will therefore have to send PATH_RESPONSE. The performance optimization is to include a PATH_CHALLENGE with that PATH_RESPONSE, not the other way around.

```suggestion
(see {{pmtud}}), and/or include a PATH_CHALLENGE with its own PATH_RESPONSE.
```

When probing a new path, an endpoint might want to ensure that its peer has an
unused connection ID available for responses. The endpoint can send
NEW_CONNECTION_ID and PATH_CHALLENGE frames in the same packet. This ensures
that an unused connection ID will be available to the peer when sending a
response.

The number of new paths an endpoint can probe is limited by the number of

> extra Connection IDs the peer has previously supplied.

Reviewer: I would add the following clarification:

```suggestion
extra Connection IDs the peer has previously supplied. Supplying a peer with enough
Connection IDs allows it to simultaneously probe multiple paths.
```

traversal needs additional synchronization mechanisms that are not provided
here.

An endpoint MAY include other frames with the PATH_CHALLENGE and PATH_RESPONSE
frames used for path validation.  In particular, an endpoint can include
PADDING with a PATH_CHALLENGE for Path Maximum Transfer Unit (PMTU) discovery
(see {{pmtud}}), and/or include a PATH_CHALLENGE with its own PATH_RESPONSE.

When probing a new path, an endpoint might want to ensure that its peer has an
unused connection ID available for responses. The endpoint can send
NEW_CONNECTION_ID and PATH_CHALLENGE frames in the same packet. This ensures
that an unused connection ID will be available to the peer when sending a
response.

The number of new paths an endpoint can probe is limited by the number of

> extra Connection IDs the peer has previously supplied. Supplying a peer with enough
> Connection IDs allows it to simultaneously probe multiple paths.

Author: ```suggestion
extra Connection IDs the peer has previously supplied. Supplying a peer with
multiple Connection IDs allows it to simultaneously probe multiple paths.
```

> The number of new paths an endpoint can probe is limited by the number of
> extra Connection IDs the peer has previously supplied. Supplying a peer with
> multiple Connection IDs allows it to simultaneously probe multiple paths.

Reviewer 3: It's important to note that multiple paths may be probed, but I don't think we want to imply that this is the only reason to supply a peer with multiple connection IDs. How about something similar to
```suggestion
An endpoint can choose to simultaneously probe multiple paths. The number
of simultaneous paths used for probes is limited by the number of extra 
Connection IDs its peer has previously supplied, since each new local address 
used for a probe requires a previously unused Connection ID.
```

usable for this connection.  Failure to validate a path does not cause the
connection to end unless there are no valid alternative paths available.

An endpoint uses a new connection ID for probes sent from a new local address;
see {{migration-linkability}} for further discussion. An endpoint that uses
a new local address needs to ensure that at least one new connection ID is
available at the peer. That can be achieved by including a NEW_CONNECTION_ID

> frame in the probe.

Reviewer 4: I initially thought this text was essential to retain, but in fact, it stems from the assumption that the peer is required to reply on the new path (and therefore with a fresh CID).  That's not a bad thing for the peer to do and therefore including the NCID frame in a probe is also worth recommending, but it's not as critical as it used to be.

Author: Re-reading this, I think I should retain the first sentence.  I'm not sure if I need to retain the second two, given this is only about probing a new path?

Reviewer 4: The first sentence is probably duplicative, but it leads into the second -- if the peer is to respond *on the same path*, they will need an extra CID, so bundling a NCID frame is a good idea.  (That's actually why NCID is allow in probing packets, IIRC.)  But on the other hand, the peer isn't required to respond on the same path any longer, so no harm done if omitted.

Reviewer 3: It might still be worth suggesting including NCID somewhere, even if not in the same place as before, I think.

Author: I restored this paragraph and made one tweak to clarify that the reason the peer needs a new CID is if you expect to receive packets on the new local address.  PTAL.

Author: And then I found the paragraph that I thought was duplicative up above in Path Validation, so I removed this paragraph and made some adjustments to the one above, most notably adding the first sentence of this paragraph.  PTAL(again)

usable for this connection.  Failure to validate a path does not cause the
connection to end unless there are no valid alternative paths available.

> An endpoint uses a new connection ID for probes sent from a new local address;	

Reviewer 5: I think this would be clearer if we say that the "endpoint" we're talking about here is a client.

Author: This is the path validation section, not the connection migration section, FYI.  I think the intent is that path validation not be client or server specific?

to a new local address without first validating the peer's address.

To establish reachability on the new path, an endpoint initiates path	
validation ({{migrate-validate}}) on the new path.  An endpoint MAY defer path	
validation until after a peer sends the next non-probing frame to its new	
address.

When migrating, the new path might not support the endpoint's current sending
rate. Therefore, the endpoint resets its congestion controller and RTT estimate,
as described in {{migration-cc}}.

The new path might not have the same ECN capability. Therefore, the endpoint
verifies ECN capability as described in {{ecn}}.

To establish reachability on the new path, an endpoint initiates path
validation ({{migrate-validate}}) on the new path.  An endpoint MAY defer path
validation until after a peer sends the next non-probing frame to its new
address.

Path validation is necessary to verify reachability of a peer on a new network
path.  Acknowledgments cannot be used for path validation as they contain
insufficient entropy and might be spoofed.  No method is provided to establish
return reachability, as endpoints independently determine reachability on each

> direction of a path.

Reviewer 2: I think that the two latter sentences are still worth keeping.  However, I would be happier if it were moved higher though.

Author: SG, moved to the third paragraph of "Path Validation {#migrate-validate}", PTAL

In response to such a packet, an endpoint MUST start sending subsequent packets
to the new peer address and MUST initiate path validation ({{migrate-validate}})

> to verify the peer's ownership of the address if it has not already.

Reviewer: This sounds backwards to me. Don't we need to validate the path *before* sending all subsequent packets on it?

How about:

```
In response to such a packet, an endpoint MUST initiate path validation ({{migrate-validate}})
to verify the peer's ownership of the address if it has not already. Once the address is
validated, the endpoint MUST start sending subsequent packets to the new peer address.
```

(sorry the GitHub suggestion UI seems to have choked on this one)

Author: I think your suggestion is not quite right, see the next paragraph which talks about the limits on sending packets to unvalidated peer addresses.

Reviewer: Thanks, I hadn't read that. I think the current text (with or without this PR) is therefore confusing. Saying that you MUST start sending then that you MAY send before validating is strange.

Author: Would it help to move the paragraph below up some?

Author: I changed it to "send all" from "start sending" because I believe that is the intent.  Does that help?

Author: now just "send"

willing to receive at the peer's current address. Thus an endpoint can migrate
to a new local address without first validating the peer's address.

To establish reachability on the new path, an endpoint initiates path
validation ({{migrate-validate}}) on the new path.  An endpoint MAY defer path
validation until after a peer sends the next non-probing frame to its new
address.

When migrating, the new path might not support the endpoint's current sending
rate. Therefore, the endpoint resets its congestion controller and RTT estimate,
as described in {{migration-cc}}.

The new path might not have the same ECN capability. Therefore, the endpoint
verifies ECN capability as described in {{ecn}}.

## Responding to Connection Migration {#migration-response}

Receiving a packet from a new peer address containing a non-probing frame
indicates that the peer has migrated to that address.

An endpoint only changes the address that it sends packets to in response to the
highest-numbered non-probing packet. This ensures that an endpoint does not send
packets to an old peer address in the case that it receives reordered packets.

> In response to such a packet, an endpoint MUST start sending subsequent packets

Author: ```suggestion
In response to such a packet, an endpoint MUST send all subsequent packets
```

To establish reachability on the new path, an endpoint initiates path
validation ({{migrate-validate}}) on the new path.  An endpoint MAY defer path
validation until after a peer sends the next non-probing frame to its new
address.

When migrating, the new path might not support the endpoint's current sending
rate. Therefore, the endpoint resets its congestion controller and RTT estimate,
as described in {{migration-cc}}.

The new path might not have the same ECN capability. Therefore, the endpoint
verifies ECN capability as described in {{ecn}}.

## Responding to Connection Migration {#migration-response}

Receiving a packet from a new peer address containing a non-probing frame
indicates that the peer has migrated to that address.

An endpoint only changes the address that it sends packets to in response to the
highest-numbered non-probing packet. This ensures that an endpoint does not send
packets to an old peer address in the case that it receives reordered packets.

In response to such a packet, an endpoint MUST send subsequent packets
to the new peer address and MUST initiate path validation ({{migrate-validate}})

> to verify the peer's ownership of the address if it has not already.

Reviewer 6: ```suggestion
to verify the peer's ownership of the address if validation is not already underway.
```

willing to receive at the peer's current address. Thus an endpoint can migrate
to a new local address without first validating the peer's address.

To establish reachability on the new path, an endpoint initiates path
validation ({{migrate-validate}}) on the new path.  An endpoint MAY defer path
validation until after a peer sends the next non-probing frame to its new

> address.

Reviewer 5: Why? Wouldn't that mean that there's a 1-RTT delay when the client switches to the new path, as the server now has to first verify the path?

Author: That's correct, it incurs a 1-RTT delay.  But it is a bit simpler to implement.

Author: This was moved from line 2324 FYI

Reviewer 3: Yeah, I'd really hope that people _not_ do that and immediately validate in response (since state usage is still governed by how many CIDs they handed out) so that we can pre-warm the validation a few RTTs before we need to migrate. Perhaps we should call out that delay as a reason to avoid deferring unless the server is out of resources?

Author: Agreed. @ekinnear if you have text in mind, would you mind writing a PR to add your preferred clarification?

highest-numbered non-probing packet. This ensures that an endpoint does not send
packets to an old peer address in the case that it receives reordered packets.

> In response to such a packet, an endpoint MUST send all subsequent packets

Reviewer 2: ```suggestion
In response to such a packet, an endpoint MUST send subsequent packets
```

"all" might require more commitment than we need here.  There might be another migration later.
